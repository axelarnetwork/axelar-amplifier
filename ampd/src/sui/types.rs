use fastcrypto::encoding::{Base58, Encoding};

use crate::sui::error::Error;

pub struct TransactionDigest([u8; 32]);

impl TransactionDigest {
    pub const fn new(digest: [u8; 32]) -> Self {
        Self(digest)
    }

    pub fn base58_encode(&self) -> String {
        Base58::encode(self.0)
    }
}

impl std::str::FromStr for TransactionDigest {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut result = [0; 32];
        result.copy_from_slice(&Base58::decode(s).map_err(|_| Error::InvalidTransactionDigest)?);

        Ok(TransactionDigest::new(result))
    }
}
