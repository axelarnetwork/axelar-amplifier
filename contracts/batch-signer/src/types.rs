use cosmwasm_schema::cw_serde;
use cosmwasm_std::{Addr, HexBinary, Uint256};
use ethabi::ethereum_types;
use std::{fmt::Display, str::FromStr};

use crate::error::ContractError;

pub type KeccackHash = [u8; 32]; // TODO: move to axelar_wasm_std, probably change to newtype with hex encoding (HexBinary/[u8;32])

pub const ID_SEPARATOR: char = '-';

pub struct Message {
    pub id: String,
    pub source_address: String,
    pub source_chain: String,
    pub destination_address: ethereum_types::Address,
    pub payload_hash: KeccackHash,
}

impl TryFrom<connection_router::types::Message> for Message {
    type Error = ContractError;

    fn try_from(msg: connection_router::types::Message) -> Result<Self, Self::Error> {
        Ok(Message {
            id: msg.id(),
            source_address: msg.source_address,
            source_chain: msg.source_domain.into(),
            destination_address: ethereum_types::Address::from_str(&msg.destination_address)
                .map_err(|_| ContractError::InvalidMessage {
                    context: "destination_address is not a valid EVM address".into(),
                })?,
            payload_hash: msg.payload_hash.as_slice().try_into().map_err(|_| {
                ContractError::InvalidMessage {
                    context: "payload_hash is not a valid keccak256 hash".into(),
                }
            })?,
        })
    }
}

// TODO: this would most likely change when other command types are supported
impl Display for Message {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "approveContractCall")
    }
}

#[cw_serde]
pub struct Proof {
    pub addresses: Vec<Addr>,
    pub weights: Vec<Uint256>,
    pub threshold: Uint256,
    pub signatures: Vec<HexBinary>,
}
