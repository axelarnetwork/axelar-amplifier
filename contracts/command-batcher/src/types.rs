use cosmwasm_schema::cw_serde;
use cosmwasm_std::{HexBinary, Uint256};

#[cw_serde]
pub struct Data {
    pub destination_chain_id: Uint256,
    pub commands_ids: Vec<[u8; 32]>,
    pub commands_types: Vec<String>,
    pub commands_params: Vec<HexBinary>,
}

#[cw_serde]
pub struct Operator {
    pub address: HexBinary,
    pub weight: Uint256,
    pub signature: Option<HexBinary>,
}

#[cw_serde]
pub struct Proof {
    pub operators: Vec<Operator>,
    pub threshold: Uint256,
}
