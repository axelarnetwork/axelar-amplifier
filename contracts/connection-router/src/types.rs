use std::str::FromStr;

use axelar_wasm_std::flagset::FlagSet;
use cosmwasm_schema::cw_serde;
use cosmwasm_std::{Addr, StdError, StdResult};
use cw_storage_plus::{Key, KeyDeserialize, Prefixer, PrimaryKey};
use flagset::flags;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::ContractError;

pub const ID_SEPARATOR: char = ':';
#[cw_serde]
pub struct MessageID {
    value: String,
}

impl FromStr for MessageID {
    type Err = ContractError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if !s.contains(ID_SEPARATOR) || s.is_empty() {
            return Err(ContractError::InvalidMessageID {});
        }
        Ok(MessageID {
            value: s.to_lowercase(),
        })
    }
}

impl From<MessageID> for String {
    fn from(d: MessageID) -> Self {
        d.value
    }
}

impl ToString for MessageID {
    fn to_string(&self) -> String {
        self.value.clone()
    }
}

impl<'a> MessageID {
    pub fn as_str(&'a self) -> &'a str {
        self.value.as_str()
    }
}

#[cw_serde]
pub struct ChainName {
    value: String,
}

impl FromStr for ChainName {
    type Err = ContractError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.contains(ID_SEPARATOR) || s.is_empty() {
            return Err(ContractError::InvalidChainName {});
        }
        Ok(ChainName {
            value: s.to_lowercase(),
        })
    }
}

impl From<ChainName> for String {
    fn from(d: ChainName) -> Self {
        d.value
    }
}

impl ToString for ChainName {
    fn to_string(&self) -> String {
        self.value.clone()
    }
}

impl<'a> PrimaryKey<'a> for ChainName {
    type Prefix = ();
    type SubPrefix = ();
    type Suffix = Self;
    type SuperSuffix = Self;

    fn key(&self) -> Vec<Key> {
        vec![Key::Ref(self.value.as_bytes())]
    }
}

impl<'a> Prefixer<'a> for ChainName {
    fn prefix(&self) -> Vec<Key> {
        vec![Key::Ref(self.value.as_bytes())]
    }
}

impl KeyDeserialize for ChainName {
    type Output = String;

    #[inline(always)]
    fn from_vec(value: Vec<u8>) -> StdResult<Self::Output> {
        String::from_utf8(value).map_err(StdError::invalid_utf8)
    }
}

#[cw_serde]
pub struct Gateway {
    pub address: Addr,
}

#[cw_serde]
pub struct ChainEndpoint {
    pub name: ChainName,
    pub gateway: Gateway,
    pub frozen_status: FlagSet<GatewayDirection>,
}

flags! {
    #[repr(u8)]
    #[derive(Deserialize, Serialize, Hash, JsonSchema)]
    pub enum GatewayDirection: u8 {
        None = 0,
        Incoming = 1,
        Outgoing = 2,
        Bidirectional = (GatewayDirection::Incoming | GatewayDirection::Outgoing).bits(),
    }
}
