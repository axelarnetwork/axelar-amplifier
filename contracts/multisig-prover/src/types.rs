use std::fmt::Display;

use axelar_wasm_std::{Participant, Snapshot};
use connection_router::state::CrossChainId;
use cosmwasm_schema::cw_serde;
use cosmwasm_std::{from_binary, HexBinary, StdResult, Uint256};
use cw_storage_plus::{Key, KeyDeserialize, PrimaryKey};
use multisig::{
    key::{PublicKey, Signature},
    worker_set::WorkerSet,
};
use sha3::{Digest, Keccak256};

use crate::encoding::{Data, Encoder};

#[cw_serde]
pub enum CommandType {
    ApproveContractCall,
    TransferOperatorship,
}

impl Display for CommandType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CommandType::ApproveContractCall => write!(f, "approveContractCall"),
            CommandType::TransferOperatorship => write!(f, "transferOperatorship"),
        }
    }
}

#[cw_serde]
pub struct Command {
    pub id: HexBinary,
    pub ty: CommandType,
    pub params: HexBinary,
}

#[cw_serde]
pub struct BatchId(HexBinary);

impl From<HexBinary> for BatchId {
    fn from(id: HexBinary) -> Self {
        Self(id)
    }
}

impl From<&[u8]> for BatchId {
    fn from(id: &[u8]) -> Self {
        Self(id.into())
    }
}

impl<'a> PrimaryKey<'a> for BatchId {
    type Prefix = ();
    type SubPrefix = ();
    type Suffix = BatchId;
    type SuperSuffix = BatchId;

    fn key(&self) -> Vec<Key> {
        vec![Key::Ref(self.0.as_slice())]
    }
}

impl KeyDeserialize for BatchId {
    type Output = BatchId;

    fn from_vec(value: Vec<u8>) -> StdResult<Self::Output> {
        Ok(from_binary(&value.into()).expect("violated invariant: BatchID is not deserializable"))
    }
}

impl BatchId {
    pub fn new(message_ids: &[CrossChainId], new_worker_set: Option<WorkerSet>) -> BatchId {
        let mut message_ids = message_ids
            .iter()
            .map(|id| id.to_string())
            .collect::<Vec<_>>();
        message_ids.sort();

        if let Some(new_worker_set) = new_worker_set {
            message_ids.push(new_worker_set.hash().to_string())
        }
        Keccak256::digest(message_ids.join(",")).as_slice().into()
    }
}

#[cw_serde]
pub struct CommandBatch {
    pub id: BatchId,
    pub message_ids: Vec<CrossChainId>,
    pub data: Data,
    pub encoder: Encoder,
}

#[cw_serde]
#[derive(Ord, PartialOrd, Eq)]
pub struct Operator {
    pub address: HexBinary,
    pub weight: Uint256,
    pub signature: Option<Signature>,
}

impl Operator {
    pub fn with_signature(self, sig: Signature) -> Operator {
        Operator {
            address: self.address,
            weight: self.weight,
            signature: Some(sig),
        }
    }
}

pub struct WorkersInfo {
    pub snapshot: Snapshot,
    pub pubkeys_by_participant: Vec<(Participant, PublicKey)>,
}
