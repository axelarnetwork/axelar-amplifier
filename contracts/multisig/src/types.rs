use std::{collections::HashMap, fmt};

use axelar_wasm_std::Snapshot;
use cosmwasm_schema::cw_serde;
use cosmwasm_std::{from_binary, Addr, HexBinary, StdResult};
use cw_storage_plus::{KeyDeserialize, PrimaryKey};

use crate::key::PublicKey;
use crate::ContractError;

#[cw_serde]
pub struct MsgToSign(HexBinary);

impl From<MsgToSign> for HexBinary {
    fn from(original: MsgToSign) -> Self {
        original.0
    }
}

impl<'a> From<&'a MsgToSign> for &'a [u8] {
    fn from(original: &'a MsgToSign) -> Self {
        original.0.as_slice()
    }
}

impl MsgToSign {
    pub fn unchecked(hex: HexBinary) -> Self {
        Self(hex)
    }
}

#[cw_serde]
pub struct KeyID {
    pub owner: Addr,
    pub subkey: String,
}

impl<'a> PrimaryKey<'a> for &KeyID {
    type Prefix = Addr;
    type SubPrefix = ();
    type Suffix = String;
    type SuperSuffix = KeyID;

    fn key(&self) -> std::vec::Vec<cw_storage_plus::Key<'_>> {
        let mut keys = self.owner.key();
        keys.extend(self.subkey.key());
        keys
    }
}

impl KeyDeserialize for KeyID {
    type Output = KeyID;

    fn from_vec(value: Vec<u8>) -> StdResult<Self::Output> {
        Ok(from_binary(&value.into()).expect("violated invariant: KeyID is not deserializable"))
    }
}

impl fmt::Display for KeyID {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({}, {})", self.owner, self.subkey)
    }
}

#[cw_serde]
pub struct Key {
    pub id: KeyID,
    pub snapshot: Snapshot,
    pub pub_keys: HashMap<String, PublicKey>,
}

#[cw_serde]
pub enum MultisigState {
    Pending,
    Completed,
}

const MESSAGE_HASH_LEN: usize = 32;

impl TryFrom<HexBinary> for MsgToSign {
    type Error = ContractError;

    fn try_from(other: HexBinary) -> Result<Self, Self::Error> {
        if other.len() != MESSAGE_HASH_LEN {
            return Err(ContractError::InvalidMessageFormat {
                reason: "Invalid input length".into(),
            });
        }

        Ok(MsgToSign::unchecked(other))
    }
}

#[cfg(test)]
mod tests {
    use cosmwasm_std::to_binary;

    use crate::test::common::test_data;

    use super::*;

    #[test]
    fn test_key_deserialize() {
        let key = KeyID {
            owner: Addr::unchecked("owner".to_string()),
            subkey: "subkey".to_string(),
        };

        let serialized = to_binary(&key).unwrap();

        assert_eq!(key, KeyID::from_vec(serialized.into()).unwrap());
    }

    #[test]
    fn test_try_from_hexbinary_to_message() {
        let hex = test_data::message();
        let message = MsgToSign::try_from(hex.clone()).unwrap();
        assert_eq!(HexBinary::from(message), hex);
    }

    #[test]
    fn test_try_from_hexbinary_to_message_fails() {
        let hex = HexBinary::from_hex("283786d844a7c4d1d424837074d0c8ec71becdcba4dd42b5307cb543a0e2c8b81c10ad541defd5ce84d2a608fc454827d0b65b4865c8192a2ea1736a5c4b72021b").unwrap();
        assert_eq!(
            MsgToSign::try_from(hex.clone()).unwrap_err(),
            ContractError::InvalidMessageFormat {
                reason: "Invalid input length".into()
            }
        );
    }
}
