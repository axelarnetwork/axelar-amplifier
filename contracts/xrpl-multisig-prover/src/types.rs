use axelar_wasm_std::VerificationStatus;
use connection_router_api::CrossChainId;
use cosmwasm_schema::cw_serde;
use cosmwasm_std::{from_json, HexBinary, Binary, StdResult, Uint256, Uint128};
use cw_storage_plus::{Key, KeyDeserialize, PrimaryKey};
use k256::ecdsa;
use k256::schnorr::signature::SignatureEncoding;
use multisig::key::Signature;
use multisig::key::PublicKey;
use ripemd::Ripemd160;
use sha2::{Digest, Sha256};

use voting_verifier::events::parse_message_id;
use crate::axelar_workers::AxelarSigner;
use crate::error::ContractError;

#[cw_serde]
pub enum TransactionStatus {
    Pending,
    Succeeded,
    FailedOnChain,
    Inconclusive,
}

#[cw_serde]
pub struct TxHash(pub HexBinary);

impl TryFrom<CrossChainId> for TxHash {
    type Error = ContractError;
    fn try_from(cc_id: CrossChainId) -> Result<Self, ContractError> {
        let (tx_id, _event_index) = parse_message_id(&cc_id.id).map_err(|_e| ContractError::InvalidMessageID(cc_id.id.to_string()))?;
        Ok(Self(HexBinary::from_hex(tx_id.to_ascii_lowercase().as_str())?))
    }
}

impl Into<HexBinary> for TxHash {
    fn into(self) -> HexBinary {
        self.0
    }
}

impl Into<TransactionStatus> for VerificationStatus {
    fn into(self) -> TransactionStatus {
        match self {
            VerificationStatus::SucceededOnChain => TransactionStatus::Succeeded,
            VerificationStatus::FailedOnChain => TransactionStatus::FailedOnChain,
            _ => TransactionStatus::Inconclusive,
        }
    }
}

#[cw_serde]
pub struct TransactionInfo {
    pub status: TransactionStatus,
    pub unsigned_contents: XRPLUnsignedTx,
    pub original_message_id: Option<CrossChainId>,
}

impl From<HexBinary> for TxHash {
    fn from(id: HexBinary) -> Self {
        Self(id)
    }
}

impl From<&[u8]> for TxHash {
    fn from(id: &[u8]) -> Self {
        Self(id.into())
    }
}

impl<'a> PrimaryKey<'a> for TxHash {
    type Prefix = ();
    type SubPrefix = ();
    type Suffix = TxHash;
    type SuperSuffix = TxHash;

    fn key(&self) -> Vec<Key> {
        vec![Key::Ref(self.0.as_slice())]
    }
}

impl KeyDeserialize for TxHash {
    type Output = TxHash;

    fn from_vec(value: Vec<u8>) -> StdResult<Self::Output> {
        Ok(from_json(&Binary::from(value)).expect("violated invariant: TxHash is not deserializable"))
    }
}

#[cw_serde]
#[derive(Ord, PartialOrd, Eq)]
pub struct Operator {
    pub address: HexBinary,
    pub weight: Uint256,
    pub signature: Option<Signature>,
}

impl Operator {
    pub fn with_signature(self, sig: Signature) -> Operator {
        Operator {
            address: self.address,
            weight: self.weight,
            signature: Some(sig),
        }
    }
}

#[cw_serde]
pub struct XRPLToken {
    pub issuer: XRPLAccountId,
    pub currency: XRPLCurrency,
}

#[cw_serde]
pub enum XRPLPaymentAmount {
    Drops(
        u64,
    ),
    Token(XRPLToken, XRPLTokenAmount),
}

#[cw_serde]
pub enum XRPLSequence {
    Plain(u32),
    Ticket(u32),
}

impl Into<u32> for XRPLSequence {
    fn into(self) -> u32 {
        match self {
            XRPLSequence::Plain(sequence) => sequence,
            XRPLSequence::Ticket(ticket) => ticket,
        }
    }
}

#[cw_serde]
pub struct XRPLSignerEntry {
    pub account: XRPLAccountId,
    pub signer_weight: u16,
}

impl From<AxelarSigner> for XRPLSignerEntry {
    fn from(signer: AxelarSigner) -> Self {
        Self {
            account: XRPLAccountId::from(&signer.pub_key),
            signer_weight: signer.weight,
        }
    }
}

#[cw_serde]
pub enum XRPLUnsignedTx {
    Payment(XRPLPaymentTx),
    SignerListSet(XRPLSignerListSetTx),
    TicketCreate(XRPLTicketCreateTx),
}

impl XRPLUnsignedTx {
    pub fn sequence(&self) -> &XRPLSequence {
        match self {
            XRPLUnsignedTx::Payment(tx) => {
                &tx.sequence
            },
            XRPLUnsignedTx::TicketCreate(tx) => {
                &tx.sequence
            },
            XRPLUnsignedTx::SignerListSet(tx) => {
                &tx.sequence
            }
        }
    }
    pub fn sequence_number_increment(&self, status: TransactionStatus) -> u32 {
        if status == TransactionStatus::Pending || status == TransactionStatus::Inconclusive {
            return 0;
        }

        match self {
            XRPLUnsignedTx::Payment(tx ) => {
                match tx.sequence {
                    XRPLSequence::Plain(_) => 1,
                    XRPLSequence::Ticket(_) => 0,
                }
            }
            XRPLUnsignedTx::SignerListSet(tx) => {
                match tx.sequence {
                    XRPLSequence::Plain(_) => 1,
                    XRPLSequence::Ticket(_) => 0,
                }
            },
            XRPLUnsignedTx::TicketCreate(tx) => {
                match status {
                    TransactionStatus::Succeeded => tx.ticket_count + 1,
                    TransactionStatus::FailedOnChain => 1,
                    TransactionStatus::Inconclusive |
                    TransactionStatus::Pending => unreachable!(),
                }
            },
        }
    }
}

#[cw_serde]
pub struct XRPLPaymentTx {
    pub account: XRPLAccountId,
    pub fee: u64,
    pub sequence: XRPLSequence,
    pub amount: XRPLPaymentAmount,
    pub destination: XRPLAccountId,
}

#[cw_serde]
pub struct XRPLSignerListSetTx {
    pub account: XRPLAccountId,
    pub fee: u64,
    pub sequence: XRPLSequence,
    pub signer_quorum: u32,
    pub signer_entries: Vec<XRPLSignerEntry>,
}

#[cw_serde]
pub struct XRPLTicketCreateTx {
    pub account: XRPLAccountId,
    pub fee: u64,
    pub sequence: XRPLSequence,
    pub ticket_count: u32,
}

#[cw_serde]
pub struct XRPLAccountId([u8; 20]);

impl XRPLAccountId {
    pub const fn to_bytes(&self) -> [u8; 20] {
        return self.0;
    }

    pub fn from_bytes(bytes: [u8; 20]) -> Self {
        Self(bytes)
    }

    pub fn to_string(&self) -> String {
        let address_type_prefix: &[u8] = &[0x00];
        let payload = [address_type_prefix, &self.to_bytes()].concat();

        let checksum_hash1 = Sha256::digest(payload.clone());
        let checksum_hash2 = Sha256::digest(checksum_hash1);
        let checksum = &checksum_hash2[0..4];

        bs58::encode([payload, checksum.to_vec()].concat())
            .with_alphabet(bs58::Alphabet::RIPPLE)
            .into_string()
    }
}

impl From<&PublicKey> for XRPLAccountId {
    fn from(pub_key: &PublicKey) -> Self {
        let public_key_hex: HexBinary = pub_key.clone().into();

        assert!(public_key_hex.len() == 33);

        let public_key_inner_hash = Sha256::digest(public_key_hex);
        let account_id = Ripemd160::digest(public_key_inner_hash);

        return XRPLAccountId(account_id.into());
    }
}

impl TryFrom<&str> for XRPLAccountId {
    type Error = ContractError;

    fn try_from(address: &str) -> Result<Self, ContractError> {
        let res = bs58::decode(address).with_alphabet(bs58::Alphabet::RIPPLE).into_vec().map_err(|_| ContractError::InvalidAddress)?;
        // .map_err(|_| ContractError::InvalidAddress)?;
        if res.len() != 25 {
            return Err(ContractError::InvalidAddress);
        }
        let mut buffer = [0u8; 20];
        buffer.copy_from_slice(&res[1..21]);
        return Ok(XRPLAccountId(buffer))
    }
}

#[cw_serde]
pub struct XRPLSigner {
    pub account: XRPLAccountId,
    pub txn_signature: HexBinary,
    pub signing_pub_key: PublicKey,
}

impl TryFrom<(multisig::msg::Signer, multisig::key::Signature)> for XRPLSigner {
    type Error = ContractError;

    fn try_from((axelar_signer, signature): (multisig::msg::Signer, multisig::key::Signature)) -> Result<Self, ContractError> {
        let txn_signature = match signature {
            multisig::key::Signature::Ecdsa(_) |
            multisig::key::Signature::EcdsaRecoverable(_) => HexBinary::from(ecdsa::Signature::to_der(
                &ecdsa::Signature::try_from(signature.clone().as_ref())
                    .map_err(|_| ContractError::FailedToEncodeSignature)?
            ).to_vec()),
            _ => unimplemented!("Unsupported signature type"),
        };

        Ok(XRPLSigner {
            account: XRPLAccountId::from(&axelar_signer.pub_key),
            signing_pub_key: axelar_signer.pub_key.clone().into(),
            txn_signature,
        })
    }
}

#[cw_serde]
pub struct XRPLSignedTransaction {
    pub unsigned_tx: XRPLUnsignedTx,
    pub signers: Vec<XRPLSigner>
}

impl XRPLSignedTransaction {
    pub fn new(unsigned_tx: XRPLUnsignedTx, signers: Vec<XRPLSigner>) -> Self {
        Self { unsigned_tx, signers }
    }
}

#[cw_serde]
pub struct XRPLCurrency(String);

impl XRPLCurrency {
    pub fn to_bytes(self) -> [u8; 20] {
        let mut buffer = [0u8; 20];
        buffer[12..15].copy_from_slice(self.to_string().as_bytes());
        buffer
    }

    // Convert the CurrencyCode to a String
    fn to_string(self) -> String {
        self.0
    }
}

const ALLOWED_CURRENCY_CHARS: &str = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789?!@#$%^&*<>(){}[]|";

impl TryFrom<String> for XRPLCurrency {
    type Error = ContractError;

    fn try_from(s: String) -> Result<XRPLCurrency, ContractError> {
        if s.len() != 3 || s == "XRP" || !s.chars().all(|c| ALLOWED_CURRENCY_CHARS.contains(c)) {
            return Err(ContractError::InvalidCurrency);
        }
        Ok(XRPLCurrency(s))
    }
}

pub const MIN_MANTISSA: u64 = 1_000_000_000_000_000;
pub const MAX_MANTISSA: u64 = 10_000_000_000_000_000 - 1;
pub const MIN_EXPONENT: i64 = -96;
pub const MAX_EXPONENT: i64 = 80;

// XRPLTokenAmount always in canonicalized XRPL mantissa-exponent format,
// such that MIN_MANTISSA <= mantissa <= MAX_MANTISSA (or equal to zero), MIN_EXPONENT <= exponent <= MAX_EXPONENT,
// In XRPL generally it can be decimal and even negative (!) but in our case that doesn't apply.
#[cw_serde]
pub struct XRPLTokenAmount {
    mantissa: u64,
    exponent: i64
}

impl XRPLTokenAmount {
    pub fn new(mantissa: u64, exponent: i64) -> Self {
        assert!(mantissa == 0 || (MIN_MANTISSA <= mantissa && mantissa <= MAX_MANTISSA && MIN_EXPONENT <= exponent && exponent <= MAX_EXPONENT));
        Self { mantissa, exponent }
    }

    pub fn to_bytes(&self) -> [u8; 8] {
        if self.mantissa == 0 {
            0x8000000000000000u64.to_be_bytes()
        } else {
            // not xrp-bit | positive bit | 8 bits exponent | 54 bits mantissa
            (0xC000000000000000u64 | ((self.exponent + 97) as u64) << 54 | self.mantissa).to_be_bytes()
        }
    }
}

pub fn canonicalize_coin_amount(amount: Uint128, decimals: u8) -> Result<XRPLTokenAmount, ContractError>{
    let (mantissa, exponent) = canonicalize_mantissa(amount, -1 * i64::from(decimals))?;
    Ok(XRPLTokenAmount::new(mantissa, exponent))
}

// always called when XRPLTokenAmount instantiated
// see https://github.com/XRPLF/xrpl-dev-portal/blob/82da0e53a8d6cdf2b94a80594541d868b4d03b94/content/_code-samples/tx-serialization/py/xrpl_num.py#L19
pub fn canonicalize_mantissa(mut mantissa: Uint128, mut exponent: i64) -> Result<(u64, i64), ContractError> {
    let ten = Uint128::from(10u128);

    while mantissa < MIN_MANTISSA.into() && exponent > MIN_EXPONENT {
        mantissa *= ten;
        exponent -= 1;
    }

    while mantissa > MAX_MANTISSA.into() && exponent > MIN_EXPONENT {
        if exponent > MAX_EXPONENT {
            return Err(ContractError::InvalidAmount { reason: "overflow".to_string() });
        }
        mantissa /= ten;
        exponent += 1;
    }

    if exponent < MIN_EXPONENT || mantissa < MIN_MANTISSA.into() {
        return Ok((0, 1));
    }

    if exponent > MAX_EXPONENT || mantissa > MAX_MANTISSA.into() {
        return Err(ContractError::InvalidAmount { reason: format!("overflow exponent {} mantissa {}", exponent, mantissa).to_string() });
    }

    let mantissa = u64::from_be_bytes(mantissa.to_be_bytes()[8..].try_into().unwrap());

    return Ok((mantissa, exponent));
}
