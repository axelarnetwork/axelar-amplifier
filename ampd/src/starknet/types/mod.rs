pub mod array_span;
pub mod byte_array;
