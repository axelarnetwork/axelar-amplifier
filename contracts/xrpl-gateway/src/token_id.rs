use interchain_token_service::TokenId;
use router_api::ChainName;
use sha3::{Digest, Keccak256};
use xrpl_types::types::{XRPLAccountId, XRPLCurrency};

const PREFIX_TOKEN_ID: &[u8] = b"its-interchain-token-id";
const PREFIX_CUSTOM_TOKEN_SALT: &[u8] = b"custom-token-salt";
const TOKEN_FACTORY_DEPLOYER: &[u8; 32] = &[0; 32];

fn token_id(salt: [u8; 32]) -> TokenId {
    let token_id: [u8; 32] = Keccak256::digest(
        [
            Keccak256::digest(PREFIX_TOKEN_ID).as_slice(),
            TOKEN_FACTORY_DEPLOYER,
            &salt,
        ]
        .concat(),
    )
    .into();
    TokenId::new(token_id)
}

fn linked_token_deploy_salt(
    chain_name_hash: [u8; 32],
    deployer: &XRPLAccountId,
    salt: [u8; 32],
) -> [u8; 32] {
    Keccak256::digest(
        [
            Keccak256::digest(PREFIX_CUSTOM_TOKEN_SALT).as_slice(),
            &chain_name_hash,
            &deployer.as_bytes(),
            &salt,
        ]
        .concat(),
    )
    .into()
}

pub fn linked_token_id(
    chain_name_hash: [u8; 32],
    deployer: &XRPLAccountId,
    salt: [u8; 32],
) -> TokenId {
    token_id(linked_token_deploy_salt(chain_name_hash, deployer, salt))
}

pub fn chain_name_hash(chain_name: ChainName) -> [u8; 32] {
    Keccak256::digest(chain_name.to_string()).into()
}

pub fn currency_hash(currency: &XRPLCurrency) -> [u8; 32] {
    Keccak256::digest(currency.as_bytes()).into()
}

#[cfg(test)]
mod test {
    use std::str::FromStr;

    use cosmwasm_std::HexBinary;
    use router_api::ChainName;
    use xrpl_types::types::{XRPLAccountId, XRPLCurrency};

    #[test]
    pub fn token_id() {
        let salt = [0u8; 32];
        let token_id = super::token_id(salt);
        goldie::assert!(token_id.to_string());
    }

    #[test]
    pub fn linked_token_deploy_salt() {
        let chain_name_hash = [0u8; 32];
        let deployer = &XRPLAccountId::from([0u8; 20]);
        let salt = [0u8; 32];

        let deploy_salt = super::linked_token_deploy_salt(chain_name_hash, deployer, salt);
        goldie::assert!(HexBinary::from(deploy_salt).to_string());
    }

    #[test]
    pub fn linked_token_id() {
        let chain_name_hash = [0u8; 32];
        let deployer = &XRPLAccountId::from([0u8; 20]);
        let salt = [0u8; 32];

        let token_id = super::linked_token_id(chain_name_hash, deployer, salt);
        goldie::assert!(token_id.to_string());
    }

    #[test]
    pub fn chain_name_hash() {
        let chain_name = ChainName::from_str("xrpl").unwrap();
        let chain_name_hash = super::chain_name_hash(chain_name);
        goldie::assert!(HexBinary::from(chain_name_hash).to_string());
    }

    #[test]
    pub fn currency_hash() {
        let currency = XRPLCurrency::XRP;
        let currency_hash = super::currency_hash(&currency);
        goldie::assert!(HexBinary::from(currency_hash).to_string());
    }
}
