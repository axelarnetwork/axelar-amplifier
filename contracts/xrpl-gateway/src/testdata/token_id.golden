726a843298d511e025ba9709b4ceff43e9986878297da57a696bdb7bbf0eff49